### ISSUE TITLE
Windows: Toolchain/SDK parameterization

### ISSUE THREAD

**Issue by woodruffw** [osquery/osquery#5344]
Right now, the Buck-based Windows build hard-codes the Visual Studio 2017 and Windows 10 SDK revisions in `./tools/buckconfigs/windows-x86_64/toolchain/vs2017_15.5.bcfg`:

```ini
[cxx]
  cpp = C:/tools/toolchains/vs2017_15.5/BuildTools/VC/Tools/MSVC/14.12.25827/bin/HostX64/x64/cl.exe
  cpp_type = windows

  cc = C:/tools/toolchains/vs2017_15.5/BuildTools/VC/Tools/MSVC/14.12.25827/bin/HostX64/x64/cl.exe
  cc_type = windows

  cxxpp = C:/tools/toolchains/vs2017_15.5/BuildTools/VC/Tools/MSVC/14.12.25827/bin/HostX64/x64/cl.exe
  cxxpp_type = windows

  cxx = C:/tools/toolchains/vs2017_15.5/BuildTools/VC/Tools/MSVC/14.12.25827/bin/HostX64/x64/cl.exe
  cxx_type = windows

  asmpp = C:/tools/toolchains/vs2017_15.5/BuildTools/VC/Tools/MSVC/14.12.25827/bin/HostX64/x64/ml64.exe
  asmpp_type = windows

  asm = C:/tools/toolchains/vs2017_15.5/BuildTools/VC/Tools/MSVC/14.12.25827/bin/HostX64/x64/ml64.exe
  asm_type = windows_ml64

  ld = C:/tools/toolchains/vs2017_15.5/BuildTools/VC/Tools/MSVC/14.12.25827/bin/HostX64/x64/link.exe
  linker_platform = windows

  ar = C:/tools/toolchains/vs2017_15.5/BuildTools/VC/Tools/MSVC/14.12.25827/bin/HostX64/x64/lib.exe
  archiver_platform = windows

[cxx_toolchain]
  cppflags = \
    /IC:/tools/toolchains/vs2017_15.5/BuildTools/VC/Tools/MSVC/14.12.25827/include \
    /IC:/tools/toolchains/vs2017_15.5/BuildTools/VC/Tools/MSVC/14.12.25827/atlmfc/include \
    /IC:/tools/toolchains/vs2017_15.5/VC/Auxiliary/VS/include \
    /IC:/tools/toolchains/vs2017_15.5/WindowsSdk/10.0.16299.91/Include/10.0.16299.0/km \
    /IC:/tools/toolchains/vs2017_15.5/WindowsSdk/10.0.16299.91/Include/10.0.16299.0/shared \
    /IC:/tools/toolchains/vs2017_15.5/WindowsSdk/10.0.16299.91/Include/10.0.16299.0/ucrt \
    /IC:/tools/toolchains/vs2017_15.5/WindowsSdk/10.0.16299.91/Include/10.0.16299.0/um \
    /IC:/tools/toolchains/vs2017_15.5/WindowsSdk/10.0.16299.91/Include/10.0.16299.0/winrt

  cflags =

  cxxppflags = \
    /IC:/tools/toolchains/vs2017_15.5/BuildTools/VC/Tools/MSVC/14.12.25827/include \
    /IC:/tools/toolchains/vs2017_15.5/BuildTools/VC/Tools/MSVC/14.12.25827/atlmfc/include \
    /IC:/tools/toolchains/vs2017_15.5/VC/Auxiliary/VS/include \
    /IC:/tools/toolchains/vs2017_15.5/WindowsSdk/10.0.16299.91/Include/10.0.16299.0/km \
    /IC:/tools/toolchains/vs2017_15.5/WindowsSdk/10.0.16299.91/Include/10.0.16299.0/shared \
    /IC:/tools/toolchains/vs2017_15.5/WindowsSdk/10.0.16299.91/Include/10.0.16299.0/ucrt \
    /IC:/tools/toolchains/vs2017_15.5/WindowsSdk/10.0.16299.91/Include/10.0.16299.0/um \
    /IC:/tools/toolchains/vs2017_15.5/WindowsSdk/10.0.16299.91/Include/10.0.16299.0/winrt

  cxxflags =

  ldflags = \
    /LIBPATH:C:/tools/toolchains/vs2017_15.5/BuildTools/VC/Tools/MSVC/14.12.25827/lib/x64 \
    /LIBPATH:C:/tools/toolchains/vs2017_15.5/WindowsSdk/10.0.16299.91/Lib/10.0.16299.0/km/x64 \
    /LIBPATH:C:/tools/toolchains/vs2017_15.5/WindowsSdk/10.0.16299.91/Lib/10.0.16299.0/ucrt/x64 \
    /LIBPATH:C:/tools/toolchains/vs2017_15.5/WindowsSdk/10.0.16299.91/Lib/10.0.16299.0/um/x64
```

These paths either need to be set up during provisioning (similarly to the original `provision.ps1` under the CMake build), or parameterized to ensure that the build finds the correct toolchain and paths for whichever versions of Visual Studio (2017+) and the Windows SDK (10+) are installed.

**Comment by fmanco** [comment]
Thanks for the report @woodruffw . Me and @muffins are actively working on this. Once our changes land the toolchain paths are going to be automatically determined.

In fact after the changes everything will be automatically determined include the host system, so the build command will look something like: `buck build @mode/select.py osquery:osqueryd`.

**Comment by woodruffw** [comment]
That's great to hear, thanks @fmanco!

**Comment by fmanco** [comment]
FYI the first part of this landed, you can now do `buck build @mode/system.py#<flavour> osquery:osqueryd` (if `#<flavour>` is omitted `release` is assumed) but it will still use hard-coded paths for Windows, which @SAlexandru is fixing now.

**Comment by fmanco** [comment]
This is implemented. Dropping the task to @SAlexandru to update the documentation.

**Comment by woodruffw** [comment]
Thanks a ton!

**Comment by mike-myers-tob** [comment]
[I updated the Windows build documentation](https://github.com/osquery/osquery/pull/5716), was there anything else needed here or is this issue closable?

**Comment by muffins** [comment]
@mike-myers-tob nah between your documentation updates and the script landed in #5623 this should be closed.
